pkg/webhooks/imagecontentpolicies/imagecontentpolicies.go

```
@@ -0,0 +1,188 @@
+	return &ImageContentPoliciesWebhook{
+		scheme: runtime.NewScheme(),
+		log:    logf.Log.WithName(WebhookName),
+	}
+}
+
+func (w *ImageContentPoliciesWebhook) Authorized(request admission.Request) admission.Response {
+	decoder, err := admission.NewDecoder(w.scheme)
+	if err != nil {
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+	icp := configv1.ImageContentPolicy{}
+	icsp := operatorv1alpha1.ImageContentSourcePolicy{}
+	allowedResp := admission.Response{
+		AdmissionResponse: admissionv1.AdmissionResponse{
+			UID:     request.UID,
+			Allowed: true,
+		},
+	}
+
+	if len(request.Object.Raw) > 0 {
+		switch request.RequestKind.Kind {
+		case "ImageContentPolicy":
+			if err := decoder.Decode(request, &icp); err != nil {
+				w.log.Error(err, "failed to render an ImageContentPolicy from request")
+				return admission.Errored(http.StatusBadRequest, err)
+			}
+
+			if !authorizeImageContentPolicy(icp) {
+				return admission.Errored(http.StatusBadRequest, fmt.Errorf(WebhookDoc))
+			}
+		case "ImageContentSourcePolicy":
+			if err := decoder.Decode(request, &icsp); err != nil {
+				w.log.Error(err, "failed to render an ImageContentSourcePolicy from request")
+				return admission.Errored(http.StatusBadRequest, err)
+			}
+
+			if !authorizeImageContentSourcePolicy(icsp) {
+				return admission.Errored(http.StatusBadRequest, fmt.Errorf(WebhookDoc))
+			}
+		}
+	}
+
+	return allowedResp
+}
+
+func (w *ImageContentPoliciesWebhook) GetURI() string {
+	return "/" + WebhookName
+}
+
+func (w *ImageContentPoliciesWebhook) Validate(request admission.Request) bool {
+	switch {
+	case request.Kind.Kind == "ImageContentPolicy":
+	case request.Kind.Kind == "ImageContentSourcePolicy":
+		return true
+	default:
+		return false
+	}
+
+	return false
+}
+
+func (w *ImageContentPoliciesWebhook) Name() string {
+	return WebhookName
+}
+
+func (w *ImageContentPoliciesWebhook) FailurePolicy() admissionregv1.FailurePolicyType {
+	// Fail-closed
```

Why are we failing closed? This is different than other hooks, as @mjlshen points out below. Please expand on why we need to fail closed so others coming to modify this hook have the context at hand.

---

Moved my PR comment into a code comment :D

---

```
	// Fail-closed because if we allow a problematic ImageContentPolicy/ImageContentSourcePolicy through,
	// it will have significant impact on the cluster. We should not modify this to fail-open unless we have
	// other specific observability and guidance to detect misconfigured ImageContentPolicy/ImageContentSourcePolicy
	// resources.
```